### PR TITLE
Automount volumes

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes partially #6549 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

**Review by rm3l — CHANGES_REQUESTED** [review]
Works great - thanks. I had a few minor comments.

File: docs/website/docs/user-guides/advanced/automounting-volumes.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    controller.devfile.io/mount-to-containers: "true"
+```
+
+By default, resources will be mounted based on the resource name:
+
+- Secrets will be mounted to `/etc/secret/<secret-name>`
+
+- Configmaps will be mounted to `/etc/config/<configmap-name>`
+
+- Persistent volume claims will be mounted to `/tmp/<pvc-name>`
```
[the remark this answers is not in the document]
Comment (feloy) [reply docs/website/docs/user-guides/advanced/automounting-volumes.md:22]: This is from the spec: https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets

File: docs/website/docs/user-guides/advanced/automounting-volumes.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+- `controller.devfile.io/mount-as`: for secrets and configmaps only, configure how the resource should be mounted to the container
+
+    - If `controller.devfile.io/mount-as: file`, the configmap/secret will be mounted as files within the mount path. This is the default behavior.
+
+    - If `controller.devfile.io/mount-as: subpath`, the keys and values in the configmap/secret will be mounted as files within the mount path using subpath volume mounts.
+
+    - If `controller.devfile.io/mount-as: env`, the keys and values in the configmap/secret will be mounted as environment variables in all containers.
+
+    When `file` is used, the configmap is mounted as a directory within the containers, erasing any files/directories already present. When `subpath` is used, each key in the configmap/secret is mounted as a subpath volume mount in the mount path, leaving existing files intact but preventing changes to the secret/configmap from propagating into the containers without a restart.
```
[the remark this answers is not in the document]
Comment (feloy) [reply docs/website/docs/user-guides/advanced/automounting-volumes.md:36]: Not sure, this paragraph is to explain the difference of behaviour between file and subpath (where other files are kept only with subpath, but mounted files are updated in container only with file).

File: docs/website/docs/user-guides/advanced/automounting-volumes.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rm3l) [reply docs/website/docs/user-guides/advanced/automounting-volumes.md:22]: Okay - thanks. It still feels weird to me, but okay.

File: docs/website/docs/user-guides/advanced/automounting-volumes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rm3l) [reply docs/website/docs/user-guides/advanced/automounting-volumes.md:36]: Ah, in this case maybe just fixing the indentation would be ok. Right now, It is at the same level as the last item and it looks like it is related to that. But maybe it's just me :-)

image

```suggestion
  When `file` is used, the configmap is mounted as a directory within the containers, erasing any files/directories already present. When `subpath` is used, each key in the configmap/secret is mounted as a subpath volume mount in the mount path, leaving existing files intact but preventing changes to the secret/configmap from propagating into the containers without a restart.
```

File: docs/website/docs/user-guides/advanced/automounting-volumes.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (feloy) [reply docs/website/docs/user-guides/advanced/automounting-volumes.md:22]: The issue can be raised on the devfile side, but I don't want odo to have a implementation different than DevWorkSpaces.

File: docs/website/docs/user-guides/advanced/automounting-volumes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (feloy) [reply docs/website/docs/user-guides/advanced/automounting-volumes.md]: Good catch, I've changed the indentation

**Review by rm3l — APPROVED** [review]
Now waiting for a consensus on labels and annotations, but LGTM so far. Nice feature - thanks!

**Comment by feloy** [comment]
/override windows-integration-test/Windows-test

```
[FAILED] [211.331 seconds]
E2E Test starting with non-empty Directory add Binding [It] should verify developer workflow of using binding as env in innerloop
/go/odo_1/tests/e2escenarios/e2e_test.go:337
```

**Comment by feloy** [comment]
/override OpenShift-Integration-tests/OpenShift-Integration-tests 

Errors on files test cleanup
```
[FAILED] [128.660 seconds]
odo devfile deploy command tests [AfterEach] deploying devfile with long-running exec when Automount volumes are present in the namespace should mount the volumes
  [AfterEach] C:/Users/Administrator.ANSIBLE-TEST-VS/3649/tests/integration/cmd_devfile_deploy_test.go:36
  [It] C:/Users/Administrator.ANSIBLE-TEST-VS/3649/tests/integration/cmd_devfile_deploy_test.go:663
```

**Comment by feloy** [comment]
/unhold


I changed the labels /annotations with the new ones

**Review by rm3l — COMMENTED** [review]
LGTM, but I quickly wanted to check how this would behave with Service Binding, which, I remembered might also alter the Pod spec to inject `envFrom` using a secret:

```sh
odo init --devfile nodejs --starter nodejs-starter --name automount-volumes-nodejs
odo add binding --service edb-cloud-native-postgresql.Cluster.postgresql.k8s.enterprisedb.io \
  --name automount-volumes-nodejs-edb-cloud-native-postgresql \
  --service-namespace shared-services \
  --bind-as-files=false
```

If I don't have any automounting configmap/secret, it works as expected:
```
$ kubectl get pod automount-volumes-nodejs-app-68d4c554b-kldjk -o yaml
[...]
    envFrom:                                                                                                                                                                                                        
    - secretRef:                                                                                                                                                                                                    
        name: automount-volumes-nodejs-edb-cloud-native-postgresql-fb57c24f
```

```
$ kubectl exec -it automount-volumes-nodejs-app-68d4c554b-kldjk -- /bin/sh -c 'env | grep CLUSTER | sort'
Alias tip: k exec -it automount-volumes-nodejs-app-68d4c554b-kldjk -- /bin/sh -c 'env | grep CLUSTER | sort'
CLUSTER_CA.CRT=-----BEGIN CERTIFICATE-----
CLUSTER_CA.KEY=-----BEGIN EC PRIVATE KEY-----
CLUSTER_CLUSTERIP=10.96.151.42
CLUSTER_DATABASE=app
CLUSTER_HOST=edb-cloud-native-postgresql-rw
CLUSTER_PASSWORD=3IxAlTt9AUeVNvPS1TVrptKCUxUMXWIkE37cQgq3orNfZcbKHX4JJOHV8pqD0xK6
CLUSTER_PGPASS=edb-cloud-native-postgresql-rw:5432:*:postgres:3IxAlTt9AUeVNvPS1TVrptKCUxUMXWIkE37cQgq3orNfZcbKHX4JJOHV8pqD0xK6
CLUSTER_PROVIDER=enterprisedb
CLUSTER_TLS.CRT=-----BEGIN CERTIFICATE-----
CLUSTER_TLS.KEY=-----BEGIN EC PRIVATE KEY-----
CLUSTER_TYPE=postgresql
CLUSTER_USERNAME=postgres
```

If I stop the previous Dev session and apply some automounting configmap/secret (I applied the resources you used in the tests of this PR), it looks like the environment variables expected from the ServiceBinding are no longer injected.

```
$ kubectl apply -f /path/to/odo/tests/examples/manifests/config-automount/
$ odo dev
$ kubectl get pod automount-volumes-nodejs-app-67f467566c-gjwxx -o yaml
[...]
    envFrom:                                                                                                                                                                                                        
    - secretRef:                                                                                                                                                                                                    
        name: automount-env-secret                                                                                                                                                                                  
    - configMapRef:                                                                                                                                                                                                 
        name: automount-env-configmap 
```

```
$ kubectl exec -it automount-volumes-nodejs-app-67f467566c-gjwxx -- /bin/sh -c 'env | grep CLUSTER | sort' 
$ 
```

Do you have the same issue? Or is that intended? Or I might have missed something too.

**Comment by rm3l** [comment]
> LGTM, but I quickly wanted to check how this would behave with Service Binding, which, I remembered might also alter the Pod spec to inject `envFrom` using a secret:
> 
> ```shell
> odo init --devfile nodejs --starter nodejs-starter --name automount-volumes-nodejs
> odo add binding --service edb-cloud-native-postgresql.Cluster.postgresql.k8s.enterprisedb.io \
>   --name automount-volumes-nodejs-edb-cloud-native-postgresql \
>   --service-namespace shared-services \
>   --bind-as-files=false
> ```
> 
> If I don't have any automounting configmap/secret, it works as expected:
> 
> ```
> $ kubectl get pod automount-volumes-nodejs-app-68d4c554b-kldjk -o yaml
> [...]
>     envFrom:                                                                                                                                                                                                        
>     - secretRef:                                                                                                                                                                                                    
>         name: automount-volumes-nodejs-edb-cloud-native-postgresql-fb57c24f
> ```
> 
> ```
> $ kubectl exec -it automount-volumes-nodejs-app-68d4c554b-kldjk -- /bin/sh -c 'env | grep CLUSTER | sort'
> Alias tip: k exec -it automount-volumes-nodejs-app-68d4c554b-kldjk -- /bin/sh -c 'env | grep CLUSTER | sort'
> CLUSTER_CA.CRT=-----BEGIN CERTIFICATE-----
> CLUSTER_CA.KEY=-----BEGIN EC PRIVATE KEY-----
> CLUSTER_CLUSTERIP=10.96.151.42
> CLUSTER_DATABASE=app
> CLUSTER_HOST=edb-cloud-native-postgresql-rw
> CLUSTER_PASSWORD=3IxAlTt9AUeVNvPS1TVrptKCUxUMXWIkE37cQgq3orNfZcbKHX4JJOHV8pqD0xK6
> CLUSTER_PGPASS=edb-cloud-native-postgresql-rw:5432:*:postgres:3IxAlTt9AUeVNvPS1TVrptKCUxUMXWIkE37cQgq3orNfZcbKHX4JJOHV8pqD0xK6
> CLUSTER_PROVIDER=enterprisedb
> CLUSTER_TLS.CRT=-----BEGIN CERTIFICATE-----
> CLUSTER_TLS.KEY=-----BEGIN EC PRIVATE KEY-----
> CLUSTER_TYPE=postgresql
> CLUSTER_USERNAME=postgres
> ```
> 
> If I stop the previous Dev session and apply some automounting configmap/secret (I applied the resources you used in the tests of this PR), it looks like the environment variables expected from the ServiceBinding are no longer injected.
> 
> ```
> $ kubectl apply -f /path/to/odo/tests/examples/manifests/config-automount/
> $ odo dev
> $ kubectl get pod automount-volumes-nodejs-app-67f467566c-gjwxx -o yaml
> [...]
>     envFrom:                                                                                                                                                                                                        
>     - secretRef:                                                                                                                                                                                                    
>         name: automount-env-secret                                                                                                                                                                                  
>     - configMapRef:                                                                                                                                                                                                 
>         name: automount-env-configmap 
> ```
> 
> ```
> $ kubectl exec -it automount-volumes-nodejs-app-67f467566c-gjwxx -- /bin/sh -c 'env | grep CLUSTER | sort' 
> $ 
> ```
> 
> Do you have the same issue? Or is that intended? Or I might have missed something too.

It works as expected when binding as volumes (`bindAsFiles` to `true`):

- With no automounting configmaps/secrets:
```
  volumes:                                                                                                                                                                                                          
  - name: odo-projects                                                                                                                                                                                              
    persistentVolumeClaim:                                                                                                                                                                                          
      claimName: odo-projects-automount-volumes-nodejs-app                                                                                                                                                          
  - emptyDir: {}                                                                                                                                                                                                    
    name: odo-shared-data                                                                                                                                                                                           
  - name: automount-volumes-nodejs-edb-cloud-native-postgresql                                                                                                                                                      
    secret:                                                                                                                                                                                                         
      defaultMode: 420                                                                                                                                                                                              
      secretName: automount-volumes-nodejs-edb-cloud-native-postgresql-caf68dfb   
```

- With automounting configmaps/secrets:
```
volumes:
[...]
  - name: auto-secret-automount-mount-path-secret                                                                                                                                                                   
    secret:                                                                                                                                                                                                         
      defaultMode: 420                                                                                                                                                                                              
      secretName: automount-mount-path-secret                                                                                                                                                                       
  - name: auto-secret-automount-subpath-secret                                                                                                                                                                      
    secret:                                                                                                                                                                                                         
      defaultMode: 420                                                                                                                                                                                              
      secretName: automount-subpath-secret                                                                                                                                                                          
  - configMap:                                                                                                                                                                                                      
      defaultMode: 420                                                                                                                                                                                              
      name: automount-default-configmap                                                                                                                                                                             
    name: auto-cm-automount-default-configmap                                                                                                                                                                       
  - configMap:                                                                                                                                                                                                      
      defaultMode: 420                                                                                                                                                                                              
      name: automount-mount-path-configmap                                                                                                                                                                          
    name: auto-cm-automount-mount-path-configmap                                                                                                                                                                    
  - configMap:                                                                                                                                                                                                      
      defaultMode: 420                                                                                                                                                                                              
      name: automount-subpath-configmap                                                                                                                                                                             
    name: auto-cm-automount-subpath-configmap                                                                                                                                                                       
  - name: automount-volumes-nodejs-edb-cloud-native-postgresql                                                                                                                                                      
    secret:                                                                                                                                                                                                         
      defaultMode: 420                                                                                                                                                                                              
      secretName: automount-volumes-nodejs-edb-cloud-native-postgresql-caf68dfb 
[...]

```

**Comment by feloy** [comment]
> If I stop the previous Dev session and apply some automounting configmap/secret (I applied the resources you used in the tests of this PR), it looks like the environment variables expected from the ServiceBinding are no longer injected.

I'm pretty sure the service binding operator is using an `Update` command to set the envfrom field. Because of using `Update` and not `Apply`, the values set by the operator are overwritten by `odo`.

```
$ kubectl get deployments.apps automount-volumes-nodejs-app -o yaml --show-managed-fields
[...]
  managedFields:
  - apiVersion: apps/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        f:template:
          f:spec:
            f:containers:
              k:{"name":"runtime"}:
                f:envFrom: {}
    manager: manager
    operation: Update
    time: "2023-04-17T09:53:51Z"
```

**Comment by rm3l** [comment]
> > If I stop the previous Dev session and apply some automounting configmap/secret (I applied the resources you used in the tests of this PR), it looks like the environment variables expected from the ServiceBinding are no longer injected.
> 
> I'm pretty sure the service binding operator is using an `Update` command to set the envfrom field. Because of using `Update` and not `Apply`, the values set by the operator are overwritten by `odo`.
> 

Just to make sure I understand - is this an issue in `odo` or SBO ?

> ```
> $ kubectl get deployments.apps automount-volumes-nodejs-app -o yaml --show-managed-fields
> [...]
>   managedFields:
>   - apiVersion: apps/v1
>     fieldsType: FieldsV1
>     fieldsV1:
>       f:spec:
>         f:template:
>           f:spec:
>             f:containers:
>               k:{"name":"runtime"}:
>                 f:envFrom: {}
>     manager: manager
>     operation: Update
>     time: "2023-04-17T09:53:51Z"
> ```
> 
> @rm3l I think it would be better to create a separate issue to track this problem. I don't think I would be able to fix it as part of thir PR

Sure anyway - let's track this in a separate issue.
Thanks.

**Review by rm3l — APPROVED** [review]

**Comment by feloy** [comment]
/override windows-integration-test/Windows-test

Flaky tests

**Comment by feloy** [comment]
> > > If I stop the previous Dev session and apply some automounting configmap/secret (I applied the resources you used in the tests of this PR), it looks like the environment variables expected from the ServiceBinding are no longer injected.
> > 
> > 
> > I'm pretty sure the service binding operator is using an `Update` command to set the envfrom field. Because of using `Update` and not `Apply`, the values set by the operator are overwritten by `odo`.
> 
> Just to make sure I understand - is this an issue in `odo` or SBO ?
> 

I think the problem is in SBO side, which should use Server-Side Apply insteof of an Update.

**Pull request merged**